Wrapped Tag and TagList

jordankoschei-okta opened the pull request:
This is a simple wrapper over `Chip` and `Stack` to add better ergonomics around tags and taglists.

I changed MUI's default `disabled` prop to `isDisabled` to be consistent with our other components.

edburyenegren-okta commented on packages/odyssey-react-mui/src/TagList.tsx:
```diff
@@ -0,0 +1,28 @@
+};
+
+const TagList = ({ children }: TagListProps) => (
+  <Stack direction="row" spacing={2}>
```
**Non-Blocking**

@KevinGhadyani-Okta this is another place we could use custom theming internally.

`direction` and `spacing` are style properties being handled here instead of `components.tsx` -- someone updating the styles there won't necessarily know they're being applied here. However, `Stack` is a generic component, so we wouldn't want to set the `defaultProps` on `MuiStack` just for this use case.

One solution would be to have a `variant` prop on `MuiStack`, similar to [what we do with MuiAlert](https://github.com/okta/odyssey/blob/develop/packages/odyssey-react-mui/src/theme/components.types.ts#L14). For ex: `variant="taglist"`. However, `MuiStack` has no `variant` prop to modify.

KevinGhadyani-Okta replied:
Ah, so you're saying to wrap `Stack` to provide a cleaner API for Odyssey internally and externally?

KevinGhadyani-Okta replied:
I agree 100%! We should be defining what these are.

We'll have to make a ticket for it.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Tag.tsx:
```diff
@@ -0,0 +1,27 @@
+  isDisabled?: boolean;
+}
+
+const Tag = ({ isClickable, isDisabled, label, ...rest }: TagProps) => (
+  <Chip label={label} clickable={isClickable} disabled={isDisabled} {...rest} />
+);
```
I'd like to avoid using `...rest` on Odyssey component going forward. If there's a prop we allow, we should allow it. The `TagProps` type doesn't even define `...rest`, so it doesn't make sense to include it as someone would have to `ts-ignore`.

KevinGhadyani-Okta replied:
**OPINION**

I would also rename `onDelete` to `onRemove`.

- "Delete" is what you do to data like deleting it from a database.
- "Remove" is something you do to UI like removing a row from a table.

When I think "delete", I think "gone for good". But you can always bring back that tag.

What do you think?

KevinGhadyani-Okta commented on packages/odyssey-storybook/src/components/odyssey-mui/Tag/Tag.stories.tsx:
```diff
@@ -59,30 +59,30 @@ const handleDelete = () => {
 
 const DefaultTemplate: Story = (args) => {
   return (
-    <Stack direction="row" spacing={2}>
-      <Chip
+    <TagList>
+      <Tag
```
Make sure we're typing the `Story` type as well: `Story<TagProps>`.

KevinGhadyani-Okta commented on packages/odyssey-storybook/src/components/odyssey-mui/Tag/Tag.stories.tsx:
```diff
@@ -11,14 +11,14 @@
 import TagMdx from "./Tag.mdx";
 
 export default {
   title: `MUI Components/Tag`,
-  component: Chip,
+  component: Tag,
   parameters: {
     docs: {
       page: TagMdx,
```
We need to make sure we're typing the default export (which is `storybookMeta` as `Meta<TagProps>`); otherwise, we're not getting any benefit from TypeScript and will run into issues where our Storybook docs don't match the actual TS.

jordankoschei-okta replied:
Only problem with this is that the story params don't line up with the `TagProps`; for example, we allow the user to specify `label1`, `label2`, `label3` so they can show multiple tags in the canvas, but obviously those aren't in `TagProps`. Is there a way around this that still gets us all the benefits of TS?

jordankoschei-okta replied:
This seems to be a common theme in our stories — we have parameters that don't line up exactly with the component props.

Should we aim to keep those aligned? Or just not worry about using the types in the stories?

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/TagList.tsx:
```diff
@@ -0,0 +1,28 @@
+  children: ReactElement<typeof Tag> | Array<ReactElement<typeof Tag>>;
+};
+
+const TagList = ({ children }: TagListProps) => (
+  <Stack component="ul" direction="row" spacing={2}>
+    {children}
+  </Stack>
+);
```
With MUI, we can't do this, but ideally, this would contain both the `ul` and `li` instead of it being separated out between two components. It's tight-coupling that I'd like to avoid if possible because it requires you know the source code of 2 components to use it.

I talk about it in this video: https://www.youtube.com/watch?v=n62Pc4KV4SM

jordankoschei-okta replied:
If we didn't have MUI constraints, what's the approach you would prefer? Make it a single component that accepts an array of tags, and renders a `<ul>` with the tag `<li>`s inside?

jordankoschei-okta commented:
I made a few changes that address the comments raised previously:

1. Tags now render as `div`s unless they're inside a TagList, in which case they render as `li`s
2. I removed the explicit `isInteractive` boolean and made it implicit; the interactive state is now shown if `onClick` is defined.
3. I modified the story controls to directly mirror the component props, so we can type them properly.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Tag.tsx:
```diff
@@ -0,0 +1,41 @@
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import { Chip } from ".";
```
```suggestion
import { Chip } from "@mui/material";
```

We should start importing all these from MUI. No more `.` imports in files that are referenced from `.` itself. Those are cyclical and cause issues in consuming apps.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Tag.tsx:
```diff
@@ -0,0 +1,41 @@
+  onDelete?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
+};
+
+const Tag = ({ isDisabled, label, onClick, onDelete }: TagProps) => {
```
What'd you think about `onRemove` vs `onDelete`?

Delete makes me think the tag's gone completely. Here, it's only removed as a potential value.

edburyenegren-okta replied:
> Delete makes me think the tag's gone completely. Here, it's only removed as a potential value.

This is true for `Autocomplete` but may not be accurate for other uses of `Tag` -- for instance, if you are editing an Application in the OIN and remove the `O365` tag.

No problem with the rename, just want to flag the dual use case.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Tag.tsx:
```diff
@@ -0,0 +1,41 @@
+export type TagProps = {
+  isDisabled?: boolean;
+  label: string;
+  onClick?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
+  onDelete?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
+};
```
```suggestion
  onClick?: ClickEventHandler<HTMLDivElement>;
  onDelete?: ClickEventHandler<HTMLDivElement>;
};
```

`ClickEventHandler` comes from the "react" library.

If this doesn't work, another safe way to do callbacks is referencing the MUI type directly:
```suggestion
  onClick?: MuiTagProps["onClick"];
  onDelete?: MuiTagProps["onDelete"];
};
```

I actually prefer this because we don't control these props, and MUI knows it's a `div`. We can't guarantee it'll always be a `<div>` that from our end.

Even if I don't like their types, this is the safest way for functions like this. Ideally, we would completely control it, but we can't always do that especially when it's relative to the underlying HTML.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Tag.tsx:
```diff
@@ -0,0 +1,41 @@
+  const { chipComponent } = useContext(TagListContext);
+
+  return (
+    <Chip
+      label={label}
+      clickable={onClick ? true : false}
+      component={chipComponent}
+      disabled={isDisabled}
+      onClick={onClick}
+      onDelete={onDelete}
+    />
```
Do we want `clickable` based on `onClick`? Doesn't this manage all interactions including deleting? For instance, you don't pass `onClick`, but you pass `onDelete`.

Now what?

What about `onClick || onDelete`?

edburyenegren-okta replied:
`onClick` and `onDelete` are different in the UI: https://cbea194.ods.dev/?path=/docs/mui-components-tag--default#clickable-tags

`clickable`/`onClick` means the tag itself is clickable and has a function

`onDelete` only applies to the ✖️ that is shown for dismissible tags

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/TagList.tsx:
```diff
@@ -0,0 +1,40 @@
+  children: ReactElement<typeof Tag> | Array<ReactElement<typeof Tag>>;
+};
+
+const TagList = ({ children }: TagListProps) => {
+  const providerValue = useMemo(
+    () => ({
+      chipComponent: "li" as ChipComponentType,
+    }),
+    []
+  );
```
Pretty sure we can type this object to avoid `as`:
```suggestion
const TagList = ({ children }: TagListProps) => {
  const providerValue = useMemo(
    () => ({
      chipComponent: "li",
    }: {
      chipComponent: ChipComponentType
    }),
    []
  );
```

KevinGhadyani-Okta replied:
I would also rename `chipComponent` to `chipElementType` or something referencing the fact that it's not a React component object.

I would assume I could do `<div>{chipComponent}</div>`, and I can, but it doesn't render a chip, it renders the text `'li'` or `'div'`.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/TagListContext.tsx:
```diff
@@ -0,0 +1,19 @@
+import { ChipProps } from "@mui/material";
+
+export type ChipComponentType = ElementType<ChipProps>;
+export const TagListContext = createContext({
+  chipComponent: "div" as ChipComponentType,
+});
```
```suggestion
export const TagListContext = createContext({
  chipComponent: "div",
}: {
  chipComponent: ChipComponentType;
});
```

Explanation _above_.

KevinGhadyani-Okta commented on packages/odyssey-storybook/src/components/odyssey-mui/Tag/Tag.stories.tsx:
```diff
@@ -10,15 +10,15 @@
 import TagMdx from "./Tag.mdx";
 
 export default {
   title: `MUI Components/Tag`,
-  component: Chip,
+  component: Tag,
   parameters: {
     docs: {
       page: TagMdx,
```
Let's do that thing where we make this:
```
const storybookMeta: Meta<TagProps> = {};
export default storybookMeta;
```

jordankoschei-okta replied:
Happy to change it, but what's the rationale of doing it this way vs the current:

```
export default {
  ...
} as Meta<TagProps>;
```

KevinGhadyani-Okta commented on packages/odyssey-storybook/src/components/odyssey-mui/Tag/Tag.stories.tsx:
```diff
@@ -29,83 +29,69 @@ export default {
 
 export const Clickable = DefaultTemplate.bind({});
 Clickable.args = {
-  clickable: true,
+  onClick: () => {
+    return true;
+  },
 };
 
 export const Deletable = DefaultTemplate.bind({});
 Deletable.args = {
-  deletable: true,
+  onDelete: () => {
+    return true;
+  },
 };
```
Storybook has an `action('')` function you can use for these. I prefer those because it shows up correctly in the canvas.

KevinGhadyani-Okta commented on packages/odyssey-storybook/src/components/odyssey-mui/Tag/Tag.stories.tsx:
```diff
@@ -29,83 +29,69 @@ export default {
 export const Clickable = DefaultTemplate.bind({});
 Clickable.args = {
-  clickable: true,
+  onClick: () => {
+    return true;
+  },
```
```suggestion
  onClick: action('onClick'),
```

Described in another post.

You probably know this, but for anyone else, this is only necessary because they're undefined in the default props.

KevinGhadyani-Okta commented on packages/odyssey-storybook/src/components/odyssey-mui/Tag/Tag.stories.tsx:
```diff
@@ -29,83 +29,69 @@ export default {
     },
-    label2: {
+    isDisabled: {
+      control: "boolean",
+      defaultValue: false,
+    },
+    onClick: {
       control: "text",
       defaultValue: null,
     },
-    label3: {
+    onDelete: {
```
Let's make sure these default values match the default values of the components so our stories match how they will use them.

When these get auto-generated normally, the default values are the default values in our component files, not for stories.

Our default values in Odyssey are both in the component files and in the `theme/components.tsx` file.

KevinGhadyani-Okta reviewed (commented):
Reviewed.
